fibers/libevent.scm

```diff
@@ -0,0 +1,215 @@
+;;;; modify it under the terms of the GNU Lesser General Public
+;;;; License as published by the Free Software Foundation; either
+;;;; version 3 of the License, or (at your option) any later version.
+;;;;
+;;;; This library is distributed in the hope that it will be useful,
+;;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;;;; Lesser General Public License for more details.
+;;;;
+;;;; You should have received a copy of the GNU Lesser General Public
+;;;; License along with this library; if not, write to the Free Software
+;;;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+;;;;
+
+(define-module (fibers events-impl)
+    #:use-module ((ice-9 binary-ports) #:select (get-u8 put-u8))
+    #:use-module (ice-9 atomic)
+    #:use-module (ice-9 control)
+    #:use-module (ice-9 match)
+    #:use-module (srfi srfi-9)
+    #:use-module (srfi srfi-9 gnu)
+    #:use-module (rnrs bytevectors)
+    #:use-module (fibers config)
+    #:export (events-impl-create
+              events-impl-destroy
+              events-impl?
+              events-impl-add!
+              events-impl-wake!
+              events-impl-fd-finalizer
+              events-impl-run
+
+              EVENTS_IMPL_READ EVENTS_IMPL_WRITE EVENTS_IMPL_CLOSED_OR_ERROR))
+
+(eval-when (eval load compile)
+  ;; When cross-compiling, the cross-compiled 'fibers-libevent.so' cannot be
+  ;; loaded by the 'guild compile' process; skip it.
+  (unless (getenv "FIBERS_CROSS_COMPILING")
+    (dynamic-call "init_libevt"
+                  (dynamic-link (extension-library "fibers-libevent")))))
+
+
+(when (defined? 'EVREAD)
+  (export EVREAD))
+(when (defined? 'EVWRITE)
+  (export EVWRITE))
+(when (defined? 'EVCLOSED)
+  (export EVCLOSED))
+
+(define (make-wake-pipe)
+  (define (set-nonblocking! port)
+    (fcntl port F_SETFL (logior O_NONBLOCK (fcntl port F_GETFL))))
+  (let ((pair (pipe)))
+    (match pair
+      ((read-pipe . write-pipe)
+       (setvbuf write-pipe 'none)
+       (set-nonblocking! read-pipe)
+       (set-nonblocking! write-pipe)
+       (values read-pipe write-pipe)))))
+
+(define-record-type <libevt>
+  (make-libevt ls added eventsv maxevents state wake-read-pipe wake-write-pipe)
+  libevt?
+  (ls libevt-ls set-libevt-ls!)
+  (added libevt-added set-libevt-added!)
+  (eventsv libevt-eventsv set-libevt-eventsv!)
+  (maxevents libevt-maxevents set-libevt-maxevents!)
+  ;; atomic box of either 'waiting, 'not-waiting or 'dead
+  (state libevt-state)
+  (wake-read-pipe libevt-wake-read-pipe)
+  (wake-write-pipe libevt-wake-write-pipe))
+
+(define-syntax fd-offset
+  (lambda (x)
+    (syntax-case x ()
+      ((_ n)
+       #`(* n #,%sizeof-struct-event)))))
+
+(define-syntax event-offset
+  (lambda (x)
+    (syntax-case x ()
+      ((_ n)
+       #`(+ (* n #,%sizeof-struct-event)
+            #,%offsetof-struct-event-ev)))))
```

Comment:
During cross-compilation, %offset-struct-event-ev is unknown, so IIUC you'll get 'unbound variable ...' during cross-compilation. When cross-compiling, I propose to do %offsetof-struct-event-ev instead of #,%offsetof-struct-event-ev.

Reply:
Done.